pipeline_dp/contribution_bounders.py

```diff
@@ -141,3 +142,73 @@ def rekey_per_privacy_id_per_partition_key(pid_pk_dict):
         return backend.map_values(
             col, aggregate_fn,
             "Apply aggregate_fn after per privacy id contribution bounding")
+
+
+class SamplingCrossPartitionContributionBounder(ContributionBounder):
+    """Bounds the contribution of privacy_id cross partitions.
+
+    It ensures that each privacy_id contributes to not more than
+    max_partitions_contributed partitions (cross-partition contribution
+    bounding), by performing sampling if needed. It is assumed that provided
+    aggregate_fn function performs per-partition contributions.
+    """
+
+    def bound_contributions(self, col, params, backend, report_generator,
+                            aggregate_fn):
+        col = backend.map_tuple(
+            col, lambda pid, pk, v: (pid, (pk, v)),
+            "Rekey to ((privacy_id), (partition_key, value))")
+
+        col = collect_values_per_partition_key_per_privacy_id(col, backend)
+        # (privacy_id, [partition_key, [value]])
+
+        # Bound cross partition contributions with sampling.
+        sample = sampling_utils.choose_from_list_without_replacement
+        sample_size = params.max_partitions_contributed
+        col = backend.map_values(col, lambda a: sample(a, sample_size))
+
+        # (privacy_id, [partition_key, [value]])
+
+        # Unnest the list per privacy id.
+        def rekey_per_privacy_id_per_partition_key(pid_pk_v_values):
+            privacy_id, partition_values = pid_pk_v_values
+
+            for partition_key, values in partition_values:
+                yield (privacy_id, partition_key), values
+
+        col = backend.flat_map(col, rekey_per_privacy_id_per_partition_key,
+                               "Unnest per privacy id")
+        # ((privacy_id, partition_key), [value])
+
+        return backend.map_values(
+            col, aggregate_fn,
+            "Apply aggregate_fn after per privacy id contribution bounding")
+
+
+def collect_values_per_partition_key_per_privacy_id(
+        col, backend: pipeline_backend.PipelineBackend):
+    """Groups per privacy id and collects values per partition key.
+
+    The output collection is a mapping from privacy_id (i.e. each privacy_id
+     from 'col' occurs exactly once) to a list [(partition_key, [values]].
+    And for any privacy_id, each partition_key for which it contributes,
```

Comment:
This sentence sounds a bit odd. Maybe something along the lines of `For any privacy_id, each partition_key it contributes to appears exactly once.` could be better?